src/rsht.c

```diff
@@ -81,29 +79,33 @@ rsht_entry *rsht_get_hash(const rsht_ht *ht, const char *key, const unsigned lon
    * this will not shrink the table; you must also periodically rebuild the hash table yourself.
    *
    * near offsets, you will see `- 1` or `+ 1`; this is for testing for the
-   * lack of an item. if an offset is 0, we've never seen the item
+   * lack of an item. if an offset is the special value, 0, we've never seen the item.
    */
 
+  rsht_entry *entry = NULL;
+
   // first, try looking up the hash
   const size_t offset = ht->buckets[hash % ht->num_buckets];
   if (offset)
-    entry = &(ht->items[offset - 1]);
-
-  if (entry) { // if we found something
-    // make sure it's the right thing
-    if (0 != strcmp(entry->key, key)) { // if it's the wrong thing
-      size_t i = 0; // find the right thing
-      while (i < ht->num_slots_used && // # items?
-          0 != strcmp(ht->items[i].key, key)) { // matching item?
+    entry = &(ht->items[offset - 1]); // since offset is 1-indexed so 0 can be reserved
+
+  if (entry) { // if that bucket has been used, make sure its entry contains the right key
+    if (0 != strcmp(entry->key, key)) { 
+      // if it's the wrong key, linearly probe for the entry with the correct key
+      size_t i = offset; // starting right after the (incorrect) entry we found
+      while (0 != strcmp(ht->items[i].key, key) && // key still doesn't match? and
+          i < ht->num_slots_used) { // still have slots to probe?
```

Comment:
Are you sure you can flip this test? 

Consider a hash table with 10 items in it, where the last item stored has a key `A`.
Assume there exists an `A'` for which `A != A'` but `hash(A) == hash(A')`.
Because the last item stored has a key `A`, the bucket for `hash(A)` points at item index 9, meaning the value stored inside it is 10 because of the 1-indexing of the offsets.
When we ask for `A'`, we find the bucket for `hash(A')`, which is the bucket for `hash(A)`. This bucket points at item index 9, meaning the value inside it is 10.
We find this entry exists at line 92, but upon comparing the keys on line 93 we notice it is not the right entry.
At line 95 we load in the offset as our first value to search. Upon first inspection, this makes complete sense, as we are looking for "the item after item 9", which is 9+1, and the offset will also be 9+1.
However, because the former item was the last item in the array, this new index now points outside the array. This can be okay, if we test to make sure that we're within the bounds of the array before doing anything else.

Line 96 of the previous code satisfied this constraint (that line performs the check, no actions are performed on the item before the check). Line 96 of the new code violates this constraint by performing an action on the item before the check.

The test must be unflipped.

Reply:
Very good catch. I have amended the commit to reflect this.